# HG changeset patch
# User Spocke <spocke@moxiecode.com>
# Date 1401108802 -7200
#      Mon May 26 14:53:22 2014 +0200
# Node ID 73fdf9307283939fd28b8463af03456cff4be7b0
# Parent  dad5bb6d55a087f356c920cd21e51d50a534b6bc
Fixed bug with paste plugin not properly removing paste bin on Safari Mac when using the cmd+shift+v keyboard command.

diff --git a/js/tinymce/plugins/paste/classes/Clipboard.js b/js/tinymce/plugins/paste/classes/Clipboard.js
--- a/js/tinymce/plugins/paste/classes/Clipboard.js
+++ b/js/tinymce/plugins/paste/classes/Clipboard.js
@@ -325,15 +325,21 @@ define("tinymce/pasteplugin/Clipboard", 
 			return mimeType in clipboardContent && clipboardContent[mimeType].length > 0;
 		}
 
+		function isKeyboardPasteEvent(e) {
+			return (VK.metaKeyPressed(e) && e.keyCode == 86) || (e.shiftKey && e.keyCode == 45);
+		}
+
 		function registerEventHandlers() {
 			editor.on('keydown', function(e) {
-				if (e.isDefaultPrevented()) {
-					return;
-				}
+				// Ctrl+V or Shift+Insert
+				if (isKeyboardPasteEvent(e) && !e.isDefaultPrevented()) {
+					keyboardPastePlainTextState = e.shiftKey && e.keyCode == 86;
 
-				// Ctrl+V or Shift+Insert
-				if ((VK.metaKeyPressed(e) && e.keyCode == 86) || (e.shiftKey && e.keyCode == 45)) {
-					keyboardPastePlainTextState = e.shiftKey && e.keyCode == 86;
+					// Edge case on Safari on Mac where it doesn't handle Cmd+Shift+V correctly
+					// it fires the keydown but no paste or keyup so we are left with a paste bin
+					if (keyboardPastePlainTextState && Env.webkit && navigator.userAgent.indexOf('Version/') != -1) {
+						return;
+					}
 
 					// Prevent undoManager keydown handler from making an undo level with the pastebin in it
 					e.stopImmediatePropagation();
@@ -353,6 +359,13 @@ define("tinymce/pasteplugin/Clipboard", 
 				}
 			});
 
+			editor.on('keyup', function(e) {
+				// Ctrl+V or Shift+Insert
+				if (isKeyboardPasteEvent(e) && !e.isDefaultPrevented()) {
+					removePasteBin();
+				}
+			});
+
 			editor.on('paste', function(e) {
 				var clipboardContent = getClipboardContent(e);
 				var isKeyBoardPaste = new Date().getTime() - keyboardPasteTimeStamp < 1000;
@@ -392,7 +405,7 @@ define("tinymce/pasteplugin/Clipboard", 
 
 					// Grab HTML from Clipboard API or paste bin as a fallback
 					if (hasContentType(clipboardContent, 'text/html')) {
-						content = Utils.trimHtml(clipboardContent['text/html']);
+						content = clipboardContent['text/html'];
 					} else {
 						content = getPasteBinHtml();
 
@@ -403,6 +416,8 @@ define("tinymce/pasteplugin/Clipboard", 
 						}
 					}
 
+					content = Utils.trimHtml(getPasteBinHtml());
+
 					// WebKit has a nice bug where it clones the paste bin if you paste from for example notepad
 					// so we need to force plain text mode in this case
 					if (pasteBinElm && pasteBinElm.firstChild && pasteBinElm.firstChild.id === 'mcepastebin') {
@@ -502,9 +517,12 @@ define("tinymce/pasteplugin/Clipboard", 
 			});
 		});
 
-		// Fix for #6504 we need to remove the paste bin on IE if the user paste in a file
-		editor.on('PreProcess', function() {
-			editor.dom.remove(editor.dom.get('mcepastebin'));
+		editor.on('BeforeAddUndo', function(e) {
+			// Remove pastebin HTML incase it should be added to an undo
+			// level for example when you paste a file on older IE
+			if (e.level.content) {
+				e.level.content = e.level.content.replace(/<div id="?mcepastebin"?[^>]+>%MCEPASTEBIN%<\/div>/gi, '');
+			}
 		});
 	};
 });
